### PR TITLE
Mpr121 added debounce and thresholds config

### PR DESCRIPTION
**Pull request by mvturnho**
## Description:

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/mpr121/mpr121_const.py [hunk 1]
```diff
@@ -0,0 +1,4 @@
+CONF_TOUCH_THRESHOLD = "touch_threshold"
+CONF_RELEASE_THRESHOLD = "release_threshold"
+CONF_TOUCH_DEBOUNCE = "touch_debounce"
+CONF_RELEASE_DEBOUNCE = "release_debounce"
```
Comment (OttoWinter) [inline esphome/components/mpr121/mpr121_const.py]: I mean you can do it like this, but you can also move these constants to `mpr121/__init__.py`.

In `binary_sensor.py` you can then import them via `from . import CONF_TOUCH_THRESHOLD, CONF_RELEASE_THRESHOLD` (don't use absolute imports, that breaks this being loaded as a custom component)

File: esphome/components/mpr121/__init__.py [hunk 2]
```diff
@@ -13,10 +15,24 @@
 MULTI_CONF = True
 CONFIG_SCHEMA = cv.Schema({
     cv.GenerateID(): cv.declare_id(MPR121Component),
+    cv.Optional(CONF_RELEASE_DEBOUNCE, default=0): cv.int_range(min=0, max=7),
+    cv.Optional(CONF_TOUCH_DEBOUNCE, default=0):  cv.All(cv.Coerce(int), cv.Range(min=0, max=7)),
```
Comment (OttoWinter) [inline esphome/components/mpr121/__init__.py]: Use `int_range` also for these other values here please.

File: esphome/components/mpr121/mpr121.cpp [hunk 3]
```diff
@@ -44,6 +44,21 @@ void MPR121Component::setup() {
   // start with first 5 bits of baseline tracking
   this->write_byte(MPR121_ECR, 0x8F);
 }
+
+void MPR121Component::set_touch_debounce(uint8_t debounce) {
+  uint8_t mask = debounce << 4;
+  this->debounce_ &= 0x0f;
+  this->debounce_ |= mask;
+  ESP_LOGD(TAG, "debounce:%02x", this->debounce_);
```
Comment (OttoWinter) [inline esphome/components/mpr121/mpr121.cpp]: No logging in setters please - and this is a verbose message at most.

File: esphome/components/mpr121/mpr121.h [hunk 4]
```diff
@@ -7,6 +7,9 @@
 namespace esphome {
 namespace mpr121 {
 
+#define DEFAULT_TOUCH_THRESHOLD 12
+#define DEFAULT_RELEASE_THRESHOLD 6
```
Comment (OttoWinter) [inline esphome/components/mpr121/mpr121.h]: These values can be removed I think (defined in .py)

File: esphome/components/mpr121/mpr121.h [hunk 5]
```diff
@@ -49,21 +52,37 @@ class MPR121Channel : public binary_sensor::BinarySensor {
...
   void process(uint16_t data) { this->publish_state(static_cast<bool>(data & (1 << this->channel_))); }
+  int get_channel() { return this->channel_; };
+  void set_touch_threshold(uint8_t touch_threshold) { this->touch_threshold_ = touch_threshold; };
+  void set_release_threshold(uint8_t release_threshold) { this->release_threshold_ = release_threshold; };
+  uint8_t get_touch_threshold() { return this->touch_threshold_; };
+  uint8_t get_release_threshold() { return this->release_threshold_; };
 
  protected:
   uint8_t channel_{0};
+  uint8_t touch_threshold_{DEFAULT_TOUCH_THRESHOLD};
```
Comment (OttoWinter) [inline esphome/components/mpr121/mpr121.h]: Ok I understand the optional threshold problem now a bit more.

Though I'm not so happy about the implementation here tbh. ESPHome has a powerful type for this: `optional`.

Change this to:

```suggestion
  optional<uint8_t> touch_threshold_{};
```

and in the `for (auto *channel : this->channels_) {` loop, use 

```cpp
channel->get_touch_threshold().value_or(this->touch_threshold_);
```

File: esphome/components/mpr121/__init__.py [hunk 6]
```diff
@@ -13,10 +15,24 @@
...
+    cv.Optional(CONF_TOUCH_THRESHOLD, default=0x06): cv.All(cv.Coerce(int),
+                                                            cv.Range(min=0x05, max=0x30)),
+    cv.Optional(CONF_RELEASE_THRESHOLD, default=0x0b): cv.All(cv.Coerce(int),
+                                                              cv.Range(min=0x05, max=0x30)),
 }).extend(cv.COMPONENT_SCHEMA).extend(i2c.i2c_device_schema(0x5A))
 
 
 def to_code(config):
     var = cg.new_Pvariable(config[CONF_ID])
+    if CONF_TOUCH_DEBOUNCE in config:
```
Comment (OttoWinter) [inline esphome/components/mpr121/__init__.py]: These if statements can be removed, the values are now always in the config (defaulted)

File: esphome/components/mpr121/mpr121.h (hunk 5 above)
Comment (mvturnho) [reply esphome/components/mpr121/mpr121.h]: I made `MPR121Component` a friend class of `MPR121Channel`, this way I could skip the getters and access the `MPR121Channel` protected threshold variables.

mpr121.h
```cpp
class MPR121Channel : public binary_sensor::BinarySensor {
  friend class MPR121Component;

 public:
  void set_channel(uint8_t channel) { channel_ = channel; }
  void process(uint16_t data) { this->publish_state(static_cast<bool>(data & (1 << this->channel_))); }
  void set_touch_threshold(uint8_t touch_threshold) { this->touch_threshold_ = touch_threshold; };
  void set_release_threshold(uint8_t release_threshold) { this->release_threshold_ = release_threshold; };

 protected:
  uint8_t channel_{0};
  optional<uint8_t> touch_threshold_{};
  optional<uint8_t> release_threshold_{};
};
```
mpr121.cpp
```cpp
  for (auto *channel : this->channels_) {
    this->write_byte(MPR121_TOUCHTH_0 + 2 * channel->channel_,
                     channel->touch_threshold_.value_or(this->touch_threshold_));
    this->write_byte(MPR121_RELEASETH_0 + 2 * channel->channel_,
                     channel->release_threshold_.value_or(this->release_threshold_));
  }
```

File: esphome/components/mpr121/binary_sensor.py [hunk 7]
```diff
@@ -11,14 +12,24 @@
...
+    cv.Optional(CONF_RELEASE_THRESHOLD): cv.int_range(min=0x05, max=0x30),
 })
 
 
 def to_code(config):
     var = cg.new_Pvariable(config[CONF_ID])
     yield binary_sensor.register_binary_sensor(var, config)
-
-    cg.add(var.set_channel(config[CONF_CHANNEL]))
+    hub = yield cg.get_variable(config[CONF_MPR121_ID])
```
Comment (OttoWinter) [inline esphome/components/mpr121/binary_sensor.py]: This line looks like a duplicate to the one below.

File: esphome/components/mpr121/binary_sensor.py [hunk 8]
```diff
@@ -11,14 +12,24 @@
...
     hub = yield cg.get_variable(config[CONF_MPR121_ID])
+    cg.add(var.set_channel(config[CONF_CHANNEL]))
+    if CONF_TOUCH_THRESHOLD in config:
+        cg.add(var.set_touch_threshold(config[CONF_TOUCH_THRESHOLD]))
+    else:
+        cg.add(var.set_touch_threshold(hub.get_touch_threshold()))
+    if CONF_RELEASE_THRESHOLD in config:
+        cg.add(var.set_release_threshold(config[CONF_RELEASE_THRESHOLD]))
+    else:
+        cg.add(var.set_release_threshold(hub.get_release_threshold()))
```
Comment (OttoWinter) [inline esphome/components/mpr121/binary_sensor.py]: Ok, now this else block can be removed.

**Review by OttoWinter — APPROVED** [review]
🎉

File: esphome/components/mpr121/__init__.py [hunk 9]
```diff
@@ -13,10 +18,18 @@
 MULTI_CONF = True
 CONFIG_SCHEMA = cv.Schema({
     cv.GenerateID(): cv.declare_id(MPR121Component),
+    cv.Optional(CONF_RELEASE_DEBOUNCE, default=0): cv.int_range(min=0, max=7),
+    cv.Optional(CONF_TOUCH_DEBOUNCE, default=0):  cv.int_range(min=0, max=7),
+    cv.Optional(CONF_TOUCH_THRESHOLD, default=0x06): cv.int_range(min=0x05, max=0x30),
+    cv.Optional(CONF_RELEASE_THRESHOLD, default=0x0b): cv.int_range(min=0x05, max=0x30),
```
Comment (OttoWinter) [inline esphome/components/mpr121/__init__.py]: The docs say that the touch threshold should be larger than the release threshold. (Also page 7 datasheet)

Why is the default release threshold here larger than the touch threshold?

**Comment by mvturnho** [comment]
@OttoWinter  also fixed the default touch / release thresholds in the docs

**Review by OttoWinter — APPROVED** [review]

**Pull request merged**